pkg/controllers/management/etcdbackup/etcdbackup.go

```diff
@@ -0,0 +1,289 @@
+		v3.BackupConditionCompleted.True(b)
+		b, err = c.backupClient.Update(b)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	kontainerDriver, err := c.KontainerDriverLister.Get("", service.RancherKubernetesEngineDriverName)
+	if err != nil {
+		return nil, err
+	}
+	bObj, err := v3.BackupConditionCompleted.Do(b, func() (runtime.Object, error) {
+		err = c.backupDriver.ETCDSave(context.Background(), cluster.Name, kontainerDriver, cluster.Spec, b.Name)
+		return b, err
+	})
+	if err != nil {
+		logrus.Errorf("[etcd-backup] failed to perform etcd backup: %v", err)
+	}
+	// need to update the object to keep the condition change.
+	return c.backupClient.Update(bObj.(*v3.EtcdBackup))
+}
+
+func (c *Controller) Remove(b *v3.EtcdBackup) (runtime.Object, error) {
+	logrus.Infof("[etcd-backup] Deleteing backup %s ", b.Name)
+	if err := c.deleteS3Snapshot(b); err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+func (c *Controller) Updated(b *v3.EtcdBackup) (runtime.Object, error) {
+	return b, nil
+}
+
+func (c *Controller) clusterBackupSync(ctx context.Context, interval time.Duration) error {
+	for range ticker.Context(ctx, interval) {
+		clusters, err := c.clusterLister.List("", labels.NewSelector())
+		if err != nil {
+			logrus.Error(fmt.Errorf("[etcd-backup] clusterBackupSync faild: %v", err))
+			return err
+		}
+		for _, cluster := range clusters {
+			logrus.Infof("[etcd-backup] Checking backups for cluster: %s", cluster.Name)
+			if err := c.doClusterBackupSync(cluster); err != nil && !apierrors.IsConflict(err) {
+				logrus.Error(fmt.Errorf("[etcd-backup] clusterBackupSync faild: %v", err))
+			}
+		}
+	}
+	return nil
+}
+
+func (c *Controller) doClusterBackupSync(cluster *v3.Cluster) error {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return nil
+	}
+
+	// not an rke cluster, we do nothing
+	if cluster.Spec.RancherKubernetesEngineConfig == nil {
+		logrus.Infof("[etcd-backup] Not an rke cluster, skipping..")
+		return nil
+	}
+	// we only work with ready clusters
+	if !v3.ClusterConditionReady.IsTrue(cluster) {
+		return nil
+	}
+	// check the backup config
+	etcdService := cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd
+	if etcdService.BackupConfig == nil {
+		// no backend backup config
+		logrus.Infof("[etcd-backup] No backup config for this cluster")
+		return nil
+	}
+
+	clusterBackups, err := c.backupLister.List(cluster.Name, labels.NewSelector())
+	if err != nil {
+		return err
+	}
+
+	// cluster has no backups, we need to kick a new one.
+	if len(clusterBackups) == 0 {
+		logrus.Infof("[etcd-backup] Cluster has no backups, creating first backup")
+		newBackup, err := c.createNewBackup(cluster)
+		if err != nil {
+			return fmt.Errorf("[etcd-backup] Backup create failed:%v", err)
+		}
+		logrus.Infof("[etcd-backup] New backup created: %s", newBackup.Name)
+		return nil
+	}
+
+	newestBackup := &v3.EtcdBackup{}
+	for _, clusterBackup := range clusterBackups {
+		if len(newestBackup.Name) == 0 {
```

Comment:
probably nit: we can avoid this check in loop by initializing newestBackup outside with first and then range from 1